From 961039030b3d3377f652a39cfdc536566735386a Mon Sep 17 00:00:00 2001
From: Robert Maynard <rmaynard@nvidia.com>
Date: Tue, 15 Nov 2022 00:49:17 -0500
Subject: [PATCH] Correct issues with CMake install rules (#331)

* Correct usage of LIBCUDACXX_TOPLEVEL_PROJECT

* Correct usage of install(DIRECTORY ... PATTERN) which matches at end of file name

* The `project` call needs to occur before any `install` commands

The install commands in cmake are only supported in project mode
and therefore need to come after the `project` call.

* Don't install any CMakeLists.txt files in libcudacxx include directory

These CMakeLists.txt files aren't needed by consumers, and therefore
shouldn't be installed.
---
 CMakeLists.txt                     | 10 +++++-----
 cmake/libcudacxxInstallRules.cmake |  6 ++++--
 2 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 5eea7447..28799729 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,6 +11,11 @@ else()
   set(LIBCUDACXX_TOPLEVEL_PROJECT OFF)
 endif()
 
+set(PACKAGE_NAME libcudacxx)
+set(PACKAGE_VERSION 11.0)
+set(PACKAGE_STRING "${PACKAGE_NAME} ${PACKAGE_VERSION}")
+project(libcudacxx NONE)
+
 include(cmake/libcudacxxInstallRules.cmake)
 
 if (NOT LIBCUDACXX_TOPLEVEL_PROJECT)
@@ -30,11 +35,6 @@ if (libcudacxx_ENABLE_CMAKE_TESTS)
   return()
 endif()
 
-set(PACKAGE_NAME libcudacxx)
-set(PACKAGE_VERSION 11.0)
-set(PACKAGE_STRING "${PACKAGE_NAME} ${PACKAGE_VERSION}")
-project(libcudacxx NONE)
-
 set(CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
 set(LLVM_PATH "${CMAKE_SOURCE_DIR}" CACHE STRING "" FORCE)
 
diff --git a/cmake/libcudacxxInstallRules.cmake b/cmake/libcudacxxInstallRules.cmake
index 446ccb50..bd92a3be 100644
--- a/cmake/libcudacxxInstallRules.cmake
+++ b/cmake/libcudacxxInstallRules.cmake
@@ -1,5 +1,5 @@
 option(libcudacxx_ENABLE_INSTALL_RULES
-  "Enable installation of libcudacxx" ${libcudacxx_TOPLEVEL_PROJECT}
+  "Enable installation of libcudacxx" ${LIBCUDACXX_TOPLEVEL_PROJECT}
 )
 
 if (NOT libcudacxx_ENABLE_INSTALL_RULES)
@@ -12,15 +12,17 @@ include(GNUInstallDirs)
 # Libcudacxx headers
 install(DIRECTORY "${libcudacxx_SOURCE_DIR}/include/cuda"
   DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
+  PATTERN CMakeLists.txt EXCLUDE
 )
 install(DIRECTORY "${libcudacxx_SOURCE_DIR}/include/nv"
   DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
+  PATTERN CMakeLists.txt EXCLUDE
 )
 
 # Libcudacxx cmake package
 install(DIRECTORY "${libcudacxx_SOURCE_DIR}/lib/cmake/libcudacxx"
   DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake"
-  PATTERN libcudacxx-header-search EXCLUDE
+  PATTERN *.cmake.in EXCLUDE
 )
 
 # Need to configure a file to store CMAKE_INSTALL_INCLUDEDIR
-- 
2.20.1

